### ISSUE TITLE
UpgradeCommand and deno compile

### ISSUE THREAD

**Issue by JohannesRudolph** [c4spar/deno-cliffy#302]
We ship some cli binaries compiled using `deno compile` because it provides a convenient single executable to download and install. One downside of this approach is that apparently `UpgradeCommand` does not work because it assumes a deno runtime instead of a compiled deno environment. 

e.g.
```
Deno.execPath(): /Users/xx/dev/mc/unipipe-service-broker/cli/bin/unipipe
  error: Unknown option "--allow-read". Did you mean option "--help"?
```

I'm not sure what the right answer is wrt. behavior. I'm thinking of using a somewhat simplified check like this right now 

```typescript
  const denoExecutable = Deno.execPath();
  const isRuntime = path.basename(denoExecutable) === "deno"; // simple and stupid, but avoids recursively invoking ourselves!

  if (!isRuntime) {
    const msg = `Upgrade is only supported when cli was installed via "deno install".`;
    program
      .command("upgrade")
      .description(msg)
      .action(() => {
        console.error(
          msg
          + `\nThis version at ${denoExecutable} appears to be a self-contained binary.\n`
          + `\nTry installing via:\n`
          + `\n\tdeno install https://raw.githubusercontent.com/meshcloud/unipipe-cli/v${VERSION}/unipipe/main.ts ${FLAGS}`
          
          );
        Deno.exit(1);
      });
  }
```

**Comment by ryo-ma** [comment]
@c4spar Do you have plans to support it?

**Comment by c4spar** [comment]
Yes i want to support this. I think we can adapt the `GithubProvider` to support downloading binaries uploaded on github. PR's are welcome.

**Comment by ryo-ma** [comment]
The function needs to read the user OS type.
It should add a prefix to the filename of the binary depending on the OS.

**Comment by c4spar** [comment]
We could do it the following:
- Detect if the cli is a binary or a script.
- Each provider should have a property which indicates if it supports binary and/or script updates.
- If no registry is provided via cli option, use the first supported provider in the options array.
- Get the github release for the requested version via github api and look for an asset which includes `Deno.build.os` and `Deno.build.arch`.
- Download the binary and install it in the target location.
- Try to auto detect the location of the binary and fallback to a provider `location` option.
- Add a global option to the upgrade command to be able to override the provider location option.
- Maybe add also a global environment variable to the upgrade command for the location which can be optionally enabled. The prefix of the env var should be also configurable.

**Comment by johnstonmatt** [comment]
My team has a solution we use for this feature, it is still a little rough around the edges but I can polish it up if it seems useful: https://github.com/polyseam/cliffy-provider-gh-releases

**Comment by ryo-ma** [comment]
@JohannesRudolph Works great, I'll try it.

**Comment by andrewthauer** [comment]
Support for private repositories with GH releases would also be a nice to have as well.

**Comment by johnstonmatt** [comment]
@andrewthauer would reading and using a token from a defined environment variable like `GITHUB_TOKEN` be enough?

**Comment by andrewthauer** [comment]
Yep.

Even better would be using an approach similar to how homebrew does with their [GitHub::API.credentials method](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/github/api.rb#L170). Try and resolve from `gh` cli, then git credential manager, followed my env vars `GITHUB_TOKEN` or `GH_TOKEN`.

Another possible option would be to use [DENO_AUTH_TOKENS](https://docs.deno.com/runtime/manual/basics/modules/private) as well.

**Comment by johnstonmatt** [comment]
@andrewthauer I added basic private repo support, it just uses those env vars for now

https://github.com/polyseam/cliffy-provider-gh-releases/releases/tag/v0.3.0
